WIP: Add text colors to print_with_color documentation (bug #15018)

LachlanGunn opened the pull request:
Currently print_with_color has no substantial documentation for the color parameter (see #15018).  I have added an automatically-generated docstring section that includes a list of possible color values.

This should keep it from getting out of date, though it does feel mildly offensive and I suspect that I may be about to inspire a new coding rule.

LachlanGunn committed:
Add text colors to print_with_color documentation.

Previously print_with_color had no substantial documentation for the
color parameter.  We have added an automatically-generated docstring
section that includes a list of possible color values.

In the course of doing so we have added a docstring to Base.text_colors
in base/client.jl.

LachlanGunn committed:
Updated print_with_color docstring in manual.

StefanKarpinski commented:
How about sorting them?

LachlanGunn commented:
Personally I think it's better not to do so programmaticaly, so that we have more flexibility to keep related ones together, e.g. if we end up with :red, and :brightred.  In any case, I'd prefer :bold and :normal not to be mixed in with the colours.

I'm happy to reorder the struct itself if you don't have a problem with the terminal codes being out of order.

StefanKarpinski commented:
Hash keys are in random order.

LachlanGunn commented:
Ah, ok.  Looking at it again carefully I see that the first two and last two are in order.  Funny coincidence.

Sorting is the way to go then, I think, though I'll try to work out if I can keep normal and bold at the end without doing anything too horrific.

tkelman commented:
A few people have suggested removing interpolation in docstrings but I think this is actually a pretty good use case.

StefanKarpinski commented:
I actually like interpolation in doc strings for this exact sort of thing.

MichaelHatherly commented:
The interpolation in `@doc_str` should probably go I think, but in plain docstrings it really is quite useful. The one quirk with it at the moment is that in `Base` where we have to use [`eval`](https://github.com/JuliaLang/julia/blob/d14a70757b585dd557c5f6ea8cbaa13323f39f26/base/docs/bootstrap.jl#L57) to insert docs after the fact any non-global vars captured by a docstring will not be defined correctly. I've got a possible solution to this which I may get around to finishing this weekend if time allows.

LachlanGunn committed:
Enhanced terminal color docstring.

Previously we just took the keys from text_colors and joined
them together.  But they came out in random order, and there
was no final "or".  We now use the format:

   <formatting>, <formatting>, <color>,
   <color>, ..., <color>, or <color>,

with colors being sorted.

LachlanGunn commented:
Ok, done.  It now goes "<formatting>*, <color>*, or <color>", where all the colors are sorted.  Here's a screenshot:
![image](https://cloud.githubusercontent.com/assets/427759/12994129/91c21f7e-d11e-11e5-929e-342a5e16d7a3.png)

StefanKarpinski commented:
Nice. Note that `join` takes an optional third argument that is an alternate join string to use between the last two items, which may be handy here.

LachlanGunn committed:
Changed text_colors docstring to use last argument of join.

LachlanGunn commented:
Ah, clever, I didn't realise.  I've changed it to use that now.

tkelman commented:
Does genstdlib need to be rerun to populate the docstring as sorted?

LachlanGunn commented:
Yes, it did need to be.  Do I need to create another pull request, or can the current one be merged from my last commit?

tkelman commented:
this was merged, so either new pr or someone can cherry pick that commit of yours and push it to master

LachlanGunn commented:
Ok, I'll make a new one and try to extricate myself from the hole I dug.

tkelman commented:
Not a big deal, it's autogenerated so would have been caught later anyway, just likely on some unrelated PR which might be a tad confusing.

LachlanGunn commented:
True.  I guess the reason it's not autogenerated is that it makes manual generation dependent upon having a julia binary?

tkelman commented:
We've tried making genstdlib always run when you do `make` so you notice the diffs sooner but we haven't figured out a way that wasn't forcing the whole system image to go through bootstrap more often than usual. Effort would probably be better spent on the markdown to html generation problem so we can take rst and sphinx out of the systen entirely. @MichaelHatherly's Lapidary.jl represents the state of the art here AIUI, I'm sure he wouldn't mind some help.

MichaelHatherly commented:
> Lapidary.jl represents the state of the art here AIUI, I'm sure he wouldn't mind some help.

Not state of the art just yet, but it'll get there :) help/bikeshedding/testing is always welcomed.
